src/coreclr/jit/emitxarch.cpp

```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
 //    true if previous instruction zeroed reg's upper 32 bits.
 //    false if it did not, or if we can't safely determine.
 //
-// Notes:
-//    Currently only looks back one instruction.
-//
-//    movsx eax, ... might seem viable but we always encode this
-//    instruction with a 64 bit destination. See TakesRexWPrefix.
-
 bool emitter::AreUpper32BitsZero(regNumber reg)
 {
+    // Only allow GPRs.
+    // If not a valid register, then return false.
+    if (!((reg >= REG_INT_FIRST) && (reg <= REG_INT_LAST)))
+        return false;
+
     // Only consider if safe
     //
     if (!emitCanPeepholeLastIns())
     {
         return false;
     }
 
-    instrDesc* id  = emitLastIns;
-    insFormat  fmt = id->idInsFmt();
+    bool result = false;
 
-    // This isn't meant to be a comprehensive check. Just look for what
-    // seems to be common.
-    switch (fmt)
-    {
-        case IF_RWR_CNS:
-        case IF_RRW_CNS:
-        case IF_RRW_SHF:
-        case IF_RWR_RRD:
-        case IF_RRW_RRD:
-        case IF_RWR_MRD:
-        case IF_RWR_SRD:
-        case IF_RWR_ARD:
+    emitPeepholeLastInstrs([&](instrDesc* id) {
+        switch (id->idInsFmt())
+        {
+            case IF_RRD:
+            case IF_RWR:
+            case IF_RRW:
 
-            // Bail if not writing to the right register
-            if (id->idReg1() != reg)
-            {
-                return false;
-            }
+            case IF_RRD_CNS:
```

Comment:
Why do we need to worry about the register read cases?

Reply:
Is this only a register read case? I was assuming anything that is `IF_R*` is something that we write to.

Reply:
We have `RRD`, `RWR`, `RRW` -- the first is a read, the second a write, and the third a read and write.